File: scripts/download-gbp-photos.mjs
import { chromium } from "playwright";
import { writeFileSync, mkdirSync } from "fs";
import { join, resolve } from "path";
import https from "https";
import http from "http";

const GBP_URL = "https://share.google/z2pczVgPxbIHSJ2he";
const OUTPUT_DIR = resolve("gbp-photos");
const DOWNLOAD_COUNT = 20;

mkdirSync(OUTPUT_DIR, { recursive: true });

function downloadFile(url, dest) {
  return new Promise((resolve, reject) => {
    const file = require("fs").createWriteStream(dest);
    const protocol = url.startsWith("https") ? https : http;
    protocol
      .get(url, (res) => {
        if (res.statusCode === 302 || res.statusCode === 301) {
          file.close();
          return downloadFile(res.headers.location, dest).then(resolve).catch(reject);
        }
        res.pipe(file);
        file.on("finish", () => file.close(resolve));
      })
      .on("error", (err) => {
        require("fs").unlink(dest, () => {});
        reject(err);
      });
  });
}

async function downloadImage(url, filepath) {
  const { default: fs } = await import("fs");
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(filepath);
    const protocol = url.startsWith("https") ? https : http;
    const request = protocol.get(url, { headers: { "User-Agent": "Mozilla/5.0" } }, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        file.close();
        downloadImage(res.headers.location, filepath).then(resolve).catch(reject);
        return;
      }
      res.pipe(file);
      file.on("finish", () => { file.close(); resolve(); });
    });
    request.on("error", (err) => { fs.unlink(filepath, () => {}); reject(err); });
  });
}

(async () => {
  const browser = await chromium.launch({ headless: false, slowMo: 200 });
  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
  });
  const page = await context.newPage();

  console.log("Opening GBP page...");
  await page.goto(GBP_URL, { waitUntil: "networkidle", timeout: 30000 });
  await page.waitForTimeout(3000);

  console.log("Current URL:", page.url());

  // Try to find and click the Photos button
  const photosSelectors = [
    'button[aria-label*="Photo"]',
    'button[aria-label*="photo"]',
    '[data-tab-index] button',
    'button:has-text("Photos")',
    'a:has-text("Photos")',
  ];

  let clickedPhotos = false;
  for (const sel of photosSelectors) {
    try {
      const el = page.locator(sel).first();
      if (await el.isVisible({ timeout: 2000 })) {
        await el.click();
        await page.waitForTimeout(2000);
        clickedPhotos = true;
        console.log("Clicked Photos tab with selector:", sel);
        break;
      }
    } catch {}
  }

  if (!clickedPhotos) {
    console.log("Could not find Photos tab — trying to find photo thumbnails directly on page...");
  }

  await page.waitForTimeout(2000);

  // Collect image URLs from thumbnails
  let imageUrls = [];
  let attempts = 0;

  while (imageUrls.length < DOWNLOAD_COUNT && attempts < 5) {
    // Scroll to load more images
    await page.evaluate(() => window.scrollBy(0, 600));
    await page.waitForTimeout(1500);

    imageUrls = await page.evaluate(() => {
      const imgs = Array.from(document.querySelectorAll("img"));
      return imgs
        .map((img) => {
          // Prefer srcset high-res, fallback to src
          if (img.srcset) {
            const parts = img.srcset.split(",").map((s) => s.trim().split(" "));
            const last = parts[parts.length - 1];
            return last[0];
          }
          return img.src;
        })
        .filter((src) => {
          if (!src || src.startsWith("data:")) return false;
          // Only grab Google content images (lh3, lh4, lh5, lh6 are GBP photos)
          return /lh[3-6]\.googleusercontent\.com/.test(src);
        })
        // Bump to higher resolution by modifying the size param
        .map((src) => src.replace(/=w\d+-h\d+.*$/, "=w1200-h900").replace(/=s\d+.*$/, "=w1200-h900"));
    });

    // Deduplicate
    imageUrls = [...new Set(imageUrls)];
    console.log(`Found ${imageUrls.length} candidate images so far...`);
    attempts++;
  }

  if (imageUrls.length === 0) {
    console.error("No Google photo URLs found. The page may require interaction.");
    await browser.close();
    process.exit(1);
  }

  const toDownload = imageUrls.slice(0, DOWNLOAD_COUNT);
  console.log(`\nDownloading ${toDownload.length} images to: ${OUTPUT_DIR}\n`);

  for (let i = 0; i < toDownload.length; i++) {
    const url = toDownload[i];
    const filename = join(OUTPUT_DIR, `photo-${String(i + 1).padStart(2, "0")}.jpg`);
    try {
      await downloadImage(url, filename);
      console.log(`  [${i + 1}/${toDownload.length}] Saved: photo-${String(i + 1).padStart(2, "0")}.jpg`);
    } catch (err) {
      console.error(`  [${i + 1}] Failed: ${err.message}`);
    }
  }

  console.log("\nDone! Check the gbp-photos folder.");
  await browser.close();
})();

File: scripts/download-gbp-photos.py
import os
import time
import urllib.request
from playwright.sync_api import sync_playwright

GBP_URL = "https://share.google/z2pczVgPxbIHSJ2he"
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "..", "gbp-photos")
DOWNLOAD_COUNT = 20

os.makedirs(OUTPUT_DIR, exist_ok=True)

def download_image(url, filepath):
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req) as response:
        with open(filepath, "wb") as f:
            f.write(response.read())

def bump_resolution(url):
    import re
    url = re.sub(r"=w\d+-h\d+.*$", "=w1200-h900", url)
    url = re.sub(r"=s\d+.*$", "=w1200-h900", url)
    return url

def collect_photo_urls(page):
    urls = page.evaluate("""() => {
        const imgs = Array.from(document.querySelectorAll('img'));
        return imgs
            .map(img => {
                if (img.srcset) {
                    const parts = img.srcset.split(',').map(s => s.trim().split(' '));
                    return parts[parts.length - 1][0];
                }
                return img.src;
            })
            .filter(src => src && !src.startsWith('data:') && /lh[3-6]\\.googleusercontent\\.com/.test(src));
    }""")
    return list(dict.fromkeys(urls))  # deduplicate, preserve order

with sync_playwright() as p:
    print("Launching Chrome...")
    browser = p.chromium.launch(channel="chrome", headless=False, slow_mo=150)
    context = browser.new_context(
        user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    )
    page = context.new_page()

    print("Opening GBP page...")
    page.goto(GBP_URL, wait_until="networkidle", timeout=30000)
    time.sleep(3)

    print("Current URL:", page.url)

    # Handle CAPTCHA / bot detection
    if "sorry" in page.url or "captcha" in page.url.lower():
        print("\n*** Google CAPTCHA detected — solve it in the browser window ***")
        print("Waiting up to 2 minutes for you to complete it...")
        for _ in range(120):
            time.sleep(1)
            if "sorry" not in page.url and "captcha" not in page.url.lower():
                break
        time.sleep(3)
        print("Current URL after CAPTCHA:", page.url)

    # Try to click the Photos tab
    photo_selectors = [
        'button[aria-label*="Photo"]',
        'button[aria-label*="photo"]',
        'button:has-text("Photos")',
        'a:has-text("Photos")',
    ]
    for sel in photo_selectors:
        try:
            el = page.locator(sel).first
            if el.is_visible(timeout=2000):
                el.click()
                time.sleep(2)
                print(f"Clicked Photos tab: {sel}")
                break
        except:
            pass

    # Scroll to load images
    image_urls = []
    for attempt in range(6):
        page.evaluate("window.scrollBy(0, 800)")
        time.sleep(1.5)
        found = collect_photo_urls(page)
        image_urls = found
        print(f"  Scroll {attempt + 1}: {len(image_urls)} candidate images found")
        if len(image_urls) >= DOWNLOAD_COUNT:
            break

    if not image_urls:
        print("No Google photo URLs found. The page may need manual interaction.")
        browser.close()
        exit(1)

    to_download = [bump_resolution(u) for u in image_urls[:DOWNLOAD_COUNT]]
    print(f"\nDownloading {len(to_download)} images to: {os.path.abspath(OUTPUT_DIR)}\n")

    for i, url in enumerate(to_download):
        filename = os.path.join(OUTPUT_DIR, f"photo-{str(i + 1).zfill(2)}.jpg")
        try:
            download_image(url, filename)
            print(f"  [{i + 1}/{len(to_download)}] Saved photo-{str(i + 1).zfill(2)}.jpg")
        except Exception as e:
            print(f"  [{i + 1}] Failed: {e}")

    print("\nDone! Check the gbp-photos folder.")
    browser.close()
